### DISCUSSION TITLE
Sorting a stream into timestamp order

### DISCUSSION THREAD

**Discussion by shubhanshu02** [bytewax/bytewax#502] [General]
## Background

There are 10 different single-partitioned Kafka topics that have data of same schema, but independent of each other. The data has a `timestamp` field which denotes the timestamp when the event was generated in IST. 

Conditions:
1. All the topics receive data from different sources in real-time.
2. All topics are single-partitioned and the data inside them is ordered as per the `timestamp` field.

## Objective

We need to read the event from all these topics in real-time and process it sequentially. 

There is a stateful operator which maintains the state of each second, but has a cache of last 20-30 seconds state (can be adjusted). To avoid out-of-order type messages, we are planning to use a `TumblingWindow` with `EventClock` to collect messages of same timestamp and then process them together. For any late messages in these windows, we can still handle them by the cache in the operator.

Note: Handling a late message has an additional cost to update all the subsequent states.

## Possible Solutions

Dataflow 1:
1. Read all topics from a `KafkaSource` in a single stream.
2. Use the `TumblingWindow` to collect the messages of same second together and then use `flat_map` to break them to stream.
3. Merge the down stream and late streams. (to include the late messages whenever they come)
4. Process the stream through the stateful operator.

Dataflow 2:
1. Read all topics in individual streams. (these individual streams are always ordered in time)
2. Collect the events of same second together in a Tumbling Window.
3. Join the tumbling windows of all streams and merge these together with all the late steams.
4. Process the stream through the stateful operator.

What of these would be a better choice? Or any better approach to achieve the same?

**Comment by davidselassie** [comment]
If your goal is to combine all of your incoming Kafka topics into a single in-timestamp-order sorted stream in the dataflow, there are a few options.

Either of your above suggestions will also work, although suggestion 1 (using `collect_window` and then `flatten` to undo the windowing) will easier to code up because it doesn't require duplicating the windowing steps for each stream. Windowing-and-then-flatten based solutions have a downside in that you have increased latency because you have to pick an arbitrary window `length` that has no effect on the output results, and downstream items will have to wait for that length. Reducing the `length` reduces this latency, but increases the overhead of the sorting.

If you need lower latency and overhead, we have a source-available commercial module with an operator `ordered` that does exactly that. https://github.com/bytewax/bytewax-ordering You are welcome to try it out now, and if you are interested in using it in production, contact sales@bytewax.io.

You can also use our low-level operator `stateful_batch` and the `Clock` primitives to create your own sorting operator, although this requires carefully manually tracking a buffer and watermarks and managing late data. You can read about the API to do so in [`StatefulBatchLogic`](https://docs.bytewax.io/stable/api/bytewax/bytewax.operators.html#bytewax.operators.StatefulBatchLogic).

---

In all cases that use a clock, you will need to set the `wait_for_system_duration` to be the maximum amount of out-of-ordered time that you might observe when reading from any two Kafka topics simultaneously. This is probably low if you are only operating on real-time data. If you need to do anything like a "backfill" and playing historical data back, if one topic is high volume with continuous observations and another topic is low volume with very sparse observations, the sparse topic input will "run ahead" to later timestamps faster (due to the consumer being able to easily read more data in a given amount of system time) and will cause the high volume items to be marked as late. This might require you to tune `wait_for_system_duration` high to support backfilling situations.

**Reply by shubhanshu02** [reply]
Can you please let me know if the following is the correct way to use the `EventClock`?

```python
def event_clock_ts(event: OrderMessage):
    timestamp = event["timestamp"]
    dt = datetime.fromtimestamp(timestamp, pytz.timezone("Asia/Kolkata"))
    return dt.astimezone(timezone.utc)
```

This in the flow:
```python
    down_stream = op.flat_map_value("window_flat_map", windowed_stream.down, lambda x: x[1])  # window id, events
    late_steam = op.map_value("late_map", windowed_stream.late, lambda x: x[1])  # window id, event
    stream = op.merge("merge_windows", down_stream, late_steam)
```

Now, when I am testing locally, I am getting some events from late stream earlier than some events from down stream in the final merged stream. 

Also, if I do not want to flatten the windows, can I still pass the two streams across the same stateful operator? Like I want to process the windows in batch and the late events after them.

**Reply by davidselassie** [reply]
Yes, that looks like a fine `ts_getter` for `EventClock`.

---

There is no global ordering of events across multiple streams in a dataflow, only within a single stream, so it is possible `merge` will interleave items. The whole point of using `collect_window` + `flatten` is to produce a single stream of ordered values. What is your thinking behind re-merging the streams afterwards because that removes the ordering guarantee?

If your downstream operator already can handle slightly out-of-timestamp-order data, then I don't know it makes sense to sort beforehand. If it can't handle out-of-timestamp-order data, then you need to treat the `late` stream with separate logic.

**Reply by shubhanshu02** [reply]
My thinking behind re-merging the two streams was - to process the late messages, we need to access the state which was maintained in the operator that processed the in-order events. So, by merging I meant to use the same state at last when all ordered messages have been processed. This is because handling late messages has a heavy cost to reiterate over the states of previous seconds and correcting all of them.
